File: apps/entity-service/src/entities/ageout.service.ts
import { Injectable, Logger, Inject, OnModuleInit } from '@nestjs/common';
import { Interval } from '@nestjs/schedule';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, DataSource } from 'typeorm';
import { ClientKafka } from '@nestjs/microservices';
import { AgeoutConfigRecord } from './ageout-config.entity';
import { KafkaTopics } from '@sentinel/common';

const DEFAULT_SEEDS = [
  { sourceType: 'ADS_B', staleThresholdMs: 60_000, ageoutThresholdMs: 300_000 },
  { sourceType: 'OPENSKY', staleThresholdMs: 60_000, ageoutThresholdMs: 300_000 },
  { sourceType: 'ADSB_LOL', staleThresholdMs: 60_000, ageoutThresholdMs: 300_000 },
  { sourceType: 'AIS', staleThresholdMs: 600_000, ageoutThresholdMs: 1_800_000 },
  { sourceType: 'CELESTRAK', staleThresholdMs: 86_400_000, ageoutThresholdMs: 604_800_000 },
  { sourceType: 'LINK16', staleThresholdMs: 30_000, ageoutThresholdMs: 120_000 },
  { sourceType: 'RADAR', staleThresholdMs: 30_000, ageoutThresholdMs: 120_000 },
  { sourceType: null, staleThresholdMs: 300_000, ageoutThresholdMs: 1_800_000 },
];

@Injectable()
export class AgeoutService implements OnModuleInit {
  private readonly logger = new Logger(AgeoutService.name);
  private isRunning = false;

  constructor(
    private readonly dataSource: DataSource,
    @InjectRepository(AgeoutConfigRecord)
    private readonly configRepo: Repository<AgeoutConfigRecord>,
    @Inject('KAFKA_CLIENT') private readonly kafkaClient: ClientKafka,
  ) {}

  async onModuleInit(): Promise<void> {
    try {
      await this.dataSource.query(
        `CREATE UNIQUE INDEX IF NOT EXISTS idx_ageout_config_global_default
         ON sentinel.feed_ageout_config ((1))
         WHERE "feedId" IS NULL AND "sourceType" IS NULL`,
      );

      const count = await this.configRepo.count();
      if (count === 0) {
        this.logger.log('Seeding default ageout thresholds');
        for (const seed of DEFAULT_SEEDS) {
          const record = new AgeoutConfigRecord();
          record.feedId = null;
          record.sourceType = seed.sourceType;
          record.staleThresholdMs = seed.staleThresholdMs;
          record.ageoutThresholdMs = seed.ageoutThresholdMs;
          await this.configRepo.save(record);
        }
      }
    } catch (error) {
      this.logger.warn(`Ageout init failed: ${error}`);
    }
  }

  @Interval(15_000)
  async processAgeout(): Promise<void> {
    if (this.isRunning) return;
    this.isRunning = true;

    try {
      const staleTransitions = await this.dataSource.query(`
        WITH candidates AS (
          SELECT e.id,
            COALESCE(
              feed_cfg."staleThresholdMs",
              source_cfg."staleThresholdMs",
              default_cfg."staleThresholdMs"
            ) AS threshold_ms
          FROM sentinel.entities e
          LEFT JOIN sentinel.feed_ageout_config feed_cfg
            ON feed_cfg."feedId" = e."feedId" AND feed_cfg."sourceType" = e.source::varchar
          LEFT JOIN sentinel.feed_ageout_config source_cfg
            ON source_cfg."feedId" IS NULL AND source_cfg."sourceType" = e.source::varchar
          LEFT JOIN sentinel.feed_ageout_config default_cfg
            ON default_cfg."feedId" IS NULL AND default_cfg."sourceType" IS NULL
          WHERE e."ageoutState" = 'LIVE' AND e.deleted = false
            AND e."lastSeenAt" IS NOT NULL
            AND EXTRACT(EPOCH FROM (NOW() - e."lastSeenAt")) * 1000 >
              COALESCE(
                feed_cfg."staleThresholdMs",
                source_cfg."staleThresholdMs",
                default_cfg."staleThresholdMs"
              )
          LIMIT 1000
        )
        UPDATE sentinel.entities SET "ageoutState" = 'STALE'
        FROM candidates c
        WHERE sentinel.entities.id = c.id
        RETURNING sentinel.entities.id, sentinel.entities."entityType",
          sentinel.entities.source, sentinel.entities."feedId",
          sentinel.entities."lastSeenAt", c.threshold_ms
      `);

      for (const row of staleTransitions) {
        this.emitAgeoutEvent(KafkaTopics.ENTITY_STALE, row, 'STALE');
      }

      const agedOutTransitions = await this.dataSource.query(`
        WITH candidates AS (
          SELECT e.id,
            COALESCE(
              feed_cfg."ageoutThresholdMs",
              source_cfg."ageoutThresholdMs",
              default_cfg."ageoutThresholdMs"
            ) AS threshold_ms
          FROM sentinel.entities e
          LEFT JOIN sentinel.feed_ageout_config feed_cfg
            ON feed_cfg."feedId" = e."feedId" AND feed_cfg."sourceType" = e.source::varchar
          LEFT JOIN sentinel.feed_ageout_config source_cfg
            ON source_cfg."feedId" IS NULL AND source_cfg."sourceType" = e.source::varchar
          LEFT JOIN sentinel.feed_ageout_config default_cfg
            ON default_cfg."feedId" IS NULL AND default_cfg."sourceType" IS NULL
          WHERE e."ageoutState" = 'STALE' AND e.deleted = false
            AND e."lastSeenAt" IS NOT NULL
            AND EXTRACT(EPOCH FROM (NOW() - e."lastSeenAt")) * 1000 >
              COALESCE(
                feed_cfg."ageoutThresholdMs",
                source_cfg."ageoutThresholdMs",
                default_cfg."ageoutThresholdMs"
              )
          LIMIT 1000
        )
        UPDATE sentinel.entities SET "ageoutState" = 'AGED_OUT'
        FROM candidates c
        WHERE sentinel.entities.id = c.id
        RETURNING sentinel.entities.id, sentinel.entities."entityType",
          sentinel.entities.source, sentinel.entities."feedId",
          sentinel.entities."lastSeenAt", c.threshold_ms
      `);

      for (const row of agedOutTransitions) {
        this.emitAgeoutEvent(KafkaTopics.ENTITY_AGED_OUT, row, 'AGED_OUT');
      }

      const total = staleTransitions.length + agedOutTransitions.length;
      if (total > 0) {
        this.logger.log(
          `Ageout cycle: ${staleTransitions.length} → STALE, ${agedOutTransitions.length} → AGED_OUT`,
        );
      }
    } catch (error) {
      this.logger.error(`Ageout processing failed: ${error}`);
    } finally {
      this.isRunning = false;
    }
  }

  private emitAgeoutEvent(
    topic: string,
    row: { id: string; entityType: string; source: string; feedId: string; lastSeenAt: Date; threshold_ms: number },
    ageoutState: string,
  ): void {
    try {
      this.kafkaClient.emit(topic, {
        key: row.id,
        value: JSON.stringify({
          entity_id: row.id,
          entity_type: row.entityType,
          source: row.source,
          feed_id: row.feedId,
          ageout_state: ageoutState,
          last_seen_at: row.lastSeenAt instanceof Date ? row.lastSeenAt.toISOString() : row.lastSeenAt,
          threshold_ms: row.threshold_ms,
          timestamp: new Date().toISOString(),
        }),
        headers: {
          'sentinel-service': 'entity-service',
          'sentinel-timestamp': new Date().toISOString(),
        },
      });
    } catch (error) {
      this.logger.error(`Failed to emit ageout event: ${error}`);
    }
  }
}

File: apps/entity-service/internal/ageout/ageout_service.go
package ageout

import (
	"context"
	"encoding/json"
	"sync"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/sentinel/go-common/kafka"
	"go.uber.org/zap"
)

// Service runs the ageout state machine on a 15-second interval.
type Service struct {
	pool     *pgxpool.Pool
	producer *kafka.Producer
	logger   *zap.Logger
	done     chan struct{}
	running  sync.Mutex
}

// NewService creates a new ageout service.
func NewService(pool *pgxpool.Pool, producer *kafka.Producer, logger *zap.Logger) *Service {
	return &Service{
		pool:     pool,
		producer: producer,
		logger:   logger,
		done:     make(chan struct{}),
	}
}

// Start begins the 15-second ageout ticker.
func (s *Service) Start() {
	go s.run()
}

func (s *Service) run() {
	ticker := time.NewTicker(15 * time.Second)
	defer ticker.Stop()

	s.logger.Info("ageout service started (15s interval)")

	for {
		select {
		case <-s.done:
			return
		case <-ticker.C:
			s.process()
		}
	}
}

func (s *Service) process() {
	if !s.running.TryLock() {
		return
	}
	defer s.running.Unlock()

	ctx := context.Background()

	staleCount := s.transitionToStale(ctx)
	agedOutCount := s.transitionToAgedOut(ctx)

	if staleCount > 0 || agedOutCount > 0 {
		s.logger.Info("ageout cycle complete",
			zap.Int("stale", staleCount),
			zap.Int("aged_out", agedOutCount),
		)
	}
}

func (s *Service) transitionToStale(ctx context.Context) int {
	rows, err := s.pool.Query(ctx, `
		WITH candidates AS (
			SELECT e.id, e."entityType", e.source, e."feedId", e."lastSeenAt"::text,
				COALESCE(
					feed_cfg."staleThresholdMs",
					source_cfg."staleThresholdMs",
					default_cfg."staleThresholdMs"
				) AS threshold_ms
			FROM sentinel.entities e
			LEFT JOIN sentinel.feed_ageout_config feed_cfg
				ON feed_cfg."feedId" = e."feedId" AND feed_cfg."sourceType" = e.source::varchar
			LEFT JOIN sentinel.feed_ageout_config source_cfg
				ON source_cfg."feedId" IS NULL AND source_cfg."sourceType" = e.source::varchar
			LEFT JOIN sentinel.feed_ageout_config default_cfg
				ON default_cfg."feedId" IS NULL AND default_cfg."sourceType" IS NULL
			WHERE e."ageoutState" = 'LIVE' AND e.deleted = false AND e."lastSeenAt" IS NOT NULL
				AND EXTRACT(EPOCH FROM (NOW() - e."lastSeenAt")) * 1000 > COALESCE(
					feed_cfg."staleThresholdMs",
					source_cfg."staleThresholdMs",
					default_cfg."staleThresholdMs"
				)
			LIMIT 1000
		)
		UPDATE sentinel.entities SET "ageoutState" = 'STALE'
		FROM candidates c WHERE sentinel.entities.id = c.id
		RETURNING c.id, c."entityType", c.source, c."feedId", c."lastSeenAt", c.threshold_ms
	`)
	if err != nil {
		s.logger.Error("failed to transition to stale", zap.Error(err))
		return 0
	}
	defer rows.Close()

	count := 0
	for rows.Next() {
		var id, entityType, source string
		var feedID, lastSeenAt *string
		var thresholdMs *int
		if err := rows.Scan(&id, &entityType, &source, &feedID, &lastSeenAt, &thresholdMs); err != nil {
			s.logger.Warn("scan stale result failed", zap.Error(err))
			continue
		}

		payload, _ := json.Marshal(map[string]interface{}{
			"entity_id":    id,
			"entity_type":  entityType,
			"source":       source,
			"feed_id":      feedID,
			"ageout_state": "STALE",
			"last_seen_at": lastSeenAt,
			"threshold_ms": thresholdMs,
			"timestamp":    time.Now().UTC().Format(time.RFC3339),
		})
		s.producer.ProduceRaw(kafka.TopicEntityStale, []byte(id), payload, nil)
		count++
	}
	return count
}

func (s *Service) transitionToAgedOut(ctx context.Context) int {
	rows, err := s.pool.Query(ctx, `
		WITH candidates AS (
			SELECT e.id, e."entityType", e.source, e."feedId", e."lastSeenAt"::text,
				COALESCE(
					feed_cfg."ageoutThresholdMs",
					source_cfg."ageoutThresholdMs",
					default_cfg."ageoutThresholdMs"
				) AS threshold_ms
			FROM sentinel.entities e
			LEFT JOIN sentinel.feed_ageout_config feed_cfg
				ON feed_cfg."feedId" = e."feedId" AND feed_cfg."sourceType" = e.source::varchar
			LEFT JOIN sentinel.feed_ageout_config source_cfg
				ON source_cfg."feedId" IS NULL AND source_cfg."sourceType" = e.source::varchar
			LEFT JOIN sentinel.feed_ageout_config default_cfg
				ON default_cfg."feedId" IS NULL AND default_cfg."sourceType" IS NULL
			WHERE e."ageoutState" = 'STALE' AND e.deleted = false AND e."lastSeenAt" IS NOT NULL
				AND EXTRACT(EPOCH FROM (NOW() - e."lastSeenAt")) * 1000 > COALESCE(
					feed_cfg."ageoutThresholdMs",
					source_cfg."ageoutThresholdMs",
					default_cfg."ageoutThresholdMs"
				)
			LIMIT 1000
		)
		UPDATE sentinel.entities SET "ageoutState" = 'AGED_OUT'
		FROM candidates c WHERE sentinel.entities.id = c.id
		RETURNING c.id, c."entityType", c.source, c."feedId", c."lastSeenAt", c.threshold_ms
	`)
	if err != nil {
		s.logger.Error("failed to transition to aged out", zap.Error(err))
		return 0
	}
	defer rows.Close()

	count := 0
	for rows.Next() {
		var id, entityType, source string
		var feedID, lastSeenAt *string
		var thresholdMs *int
		if err := rows.Scan(&id, &entityType, &source, &feedID, &lastSeenAt, &thresholdMs); err != nil {
			s.logger.Warn("scan aged out result failed", zap.Error(err))
			continue
		}

		payload, _ := json.Marshal(map[string]interface{}{
			"entity_id":    id,
			"entity_type":  entityType,
			"source":       source,
			"feed_id":      feedID,
			"ageout_state": "AGED_OUT",
			"last_seen_at": lastSeenAt,
			"threshold_ms": thresholdMs,
			"timestamp":    time.Now().UTC().Format(time.RFC3339),
		})
		s.producer.ProduceRaw(kafka.TopicEntityAgedOut, []byte(id), payload, nil)
		count++
	}
	return count
}

// Stop stops the ageout service.
func (s *Service) Stop() {
	close(s.done)
}
